README
List-MergeSorted-XS 1.05
========================

This module takes a set of presorted lists and returns the sorted union of
those lists.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

COPYRIGHT AND LICENCE

Written by Adam Thomason, <athomason@cpan.org>

Copyright (C) 2011 by Say Media Inc <cpan@saymedia.com>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.6 or,
at your option, any later version of Perl 5 you may have available.


Changes
Revision history for Perl extension List::MergeSorted::XS

1.05  Thu Apr 21 12:47:54 PDT 2011
    - use hash of seen unique values instead of comparing to last to fix corner
      case where duplicated elements that compare the same but have different
      unique values may not be deduped correctly
    - fix use of forced-algorithm in test
1.04  Tue Mar 15 00:37:22 PDT 2011
    - use merge's parameter hash instead of package variable for forcing algorithm
    - pod improvements
1.03  Tue Dec 14 13:23:24 PST 2010
    - fix incorrectly spelled instances of uniq_cb
1.02  Tue Dec 14 11:25:46 PST 2010
    - fix mis-merged code in _merge
    - speed up auto codepath by counting elements in XS and not using division
1.01  Mon Dec 13 21:29:18 PST 2010
    - add support for deduplication of elements
    - rename things and rejigger internals
1.00  Wed Dec  8 13:50:07 PST 2010
	- initial version


list.c
#include "EXTERN.h"
#include "perl.h"
#include "XSUB.h"

#include "list.h"

lmsxs_ll_ent*
lmsxs_ll_make_ent(const char* key, SV* sv, IV list_num, IV list_idx)
{
    lmsxs_ll_ent* ent;
    Newxz(ent, 1, lmsxs_ll_ent);
    ent->key = key;
    ent->sv = sv;
    ent->list_num = list_num;
    ent->list_idx = list_idx;
    ent->next = NULL;
    return ent;
}

void
lmsxs_ll_free_ent(lmsxs_ll_ent* ent)
{
    Safefree(ent);
}

void
lmsxs_ll_insert_ent(lmsxs_ll_ent** list, lmsxs_ll_ent* new_ent)
{
    lmsxs_ll_ent* cur;
    lmsxs_ll_ent** ptr_to_cur;

    if (!*list) {
        /* list is empty */
        *list = new_ent;
        return;
    }

    for (
        ptr_to_cur = list, cur = *list;
        cur;
        ptr_to_cur = &((*ptr_to_cur)->next), cur = cur->next
    ) {
        if (strcmp(new_ent->key, cur->key) < 0) {
            new_ent->next = cur;
            *ptr_to_cur = new_ent;
            return;
        }
    }

    /* insert at end */
    *ptr_to_cur = new_ent;
}

lmsxs_ll_ent*
lmsxs_ll_pop_ent(lmsxs_ll_ent** list)
{
    lmsxs_ll_ent* first = *list;
    if (!first)
        return NULL;
    *list = first->next;
    first->next = NULL;
    return first;
}


bench.pl
#!/usr/bin/perl

use strict;
use warnings;

use Benchmark;
use Getopt::Long;

use List::MergeSorted::XS 'merge';
use List::Util qw( min max );

GetOptions(
    'iters=i'       => \(my $time_count = -5),
    'elements=i'    => \(my $elements   = 10),
    'fixed-total!'  => \(my $const_total_elems = 0), # otherwise, constant elems per list
);

my $quit = 0;
$SIG{INT} = sub { exit if $quit++; print "Interrupting; CTRL-C again to quit\n" };

my @num_lists = (1, 2, 3, 5, 10, 50, 100, 500, 1000, 5000, 10000);
my @limits = (1, 5, 10, 25, 50, 75, 100);
my %overall_timings;

srand 999;

OUTER:
for my $keyed (0, 1) {
    for my $limit_percent (@limits) {
        for my $num_lists (@num_lists) {
            my $list_size = $const_total_elems ? max(int($elements / $num_lists), 1) : $elements;

            my %opts;
            $opts{key_cb} = sub { $_[0][0] } if $keyed;
            $opts{limit} = max(1, int($limit_percent / 100 * $num_lists * $list_size)) if $limit_percent;

            warn sprintf "keyed = %s; # lists = %d; list size = %d; limit = %s\n",
                $keyed ? 'on' : 'off', $num_lists, $list_size, defined $opts{limit} ? $opts{limit} : 'none';

            my @lists;
            for (1 .. $num_lists) {
                my @list = sort { $a <=> $b } map { int rand 1000000 } 1 .. $list_size;
                @list = map { [$_ => $_] } @list if $keyed;
                push @lists, \@list;
            }

            my %methods = (
                auto    => undef,
                linear  => List::MergeSorted::XS::PRIO_LINEAR,
                fib     => List::MergeSorted::XS::PRIO_FIB,
                sort    => List::MergeSorted::XS::SORT,
            );
            for my $method (sort keys %methods) {
                local $List::MergeSorted::XS::MERGE_METHOD = $methods{$method};
                my $bench = timethis($time_count, sub { merge(\@lists, %opts) }, $method);
                $overall_timings{$keyed}{$limit_percent}{$method}{$num_lists} = $bench->[5] / $bench->[1]; # iterations over user time
                last OUTER if $quit;
            }
        }
    }
}

#use Data::Dumper; print Data::Dumper->Dump(\%overall_timings);

printf "%7s %5s %12s", 'keyed', 'limit', 'method';
printf " %9d", $_ for @num_lists;
print "\n";

for my $keyed (0, 1) {
    for my $limit_percent (@limits) {
        for my $method (sort keys %{ $overall_timings{$keyed}{$limit_percent} }) {
            printf "%7s %5s %12s", $keyed, $limit_percent ? $limit_percent . '%' : 'none', $method;
            printf " %9.1f", $overall_timings{$keyed}{$limit_percent}{$method}{$_} || 0 for @num_lists;
            print "\n";
        }
    }
}


Makefile.PL
use 5.008;
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'List::MergeSorted::XS',
    VERSION_FROM      => 'XS.pm', # finds $VERSION
    PREREQ_PM         => {}, # e.g., Module::Name => 1.1
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'XS.pm', # retrieve abstract from module
       AUTHOR         => 'Adam Thomason <athomason@cpan.org>') : ()),
    LIBS              => [''], # e.g., '-lm'
    DEFINE            => '', # e.g., '-DHAVE_SOMETHING'
    INC               => '-I.', # e.g., '-I. -I/usr/include/other'
    C                 => ['fib.c', 'list.c'],
    OBJECT            => "XS.o fib.o list.o",
);
if  (eval {require ExtUtils::Constant; 1}) {
  # If you edit these definitions to change the constants used by this module,
  # you will need to use the generated const-c.inc and const-xs.inc
  # files to replace their "fallback" counterparts before distributing your
  # changes.
  my @names = (qw());
  ExtUtils::Constant::WriteConstants(
                                     NAME         => 'List::MergeSorted::XS',
                                     NAMES        => \@names,
                                     DEFAULT_TYPE => 'IV',
                                     C_FILE       => 'const-c.inc',
                                     XS_FILE      => 'const-xs.inc',
                                  );

}
else {
  use File::Copy;
  use File::Spec;
  foreach my $file ('const-c.inc', 'const-xs.inc') {
    my $fallback = File::Spec->catfile('fallback', $file);
    copy ($fallback, $file) or die "Can't copy $fallback to $file: $!";
  }
}


gen-XSxs.pl
#!/usr/bin/perl

# XS code generator for priority-queue implementations

use strict;
use warnings;

use Template;

print <DATA>;

my %impls = (
    linear => {
        type    => 'lmsxs_ll_ent',
        insert  => 'lmsxs_ll_insert_ent(&heads, ent);',
        pop     => 'lmsxs_ll_pop_ent(&heads)',
        alloc   => 'lmsxs_ll_make_ent(key, el, n, 0)',
        free    => 'lmsxs_ll_free_ent',
        more    => 'heads',
    },
    fib => {
        type    => 'lmsxs_prio_ent',
        insert  => 'fh_insertkey(heads, key, ent);',
        pop     => '(lmsxs_prio_ent*) fh_extractmin(heads)',
        alloc   => 'lmsxs_make_ent(el, n, 0)',
        free    => 'lmsxs_free_ent',
        more    => '!fh_empty(heads)',
    },
);

for my $impl (sort keys %impls) {
    for my $keyed (0, 1) {
        for my $dedupe (0, 1) {
            my %vars = (
                %{ $impls{$impl} },
                impl   => $impl,
                keyed  => $keyed,
                dedupe => $dedupe,
            );

            $vars{name} = "l_ms_xs_merge_";
            $vars{name} .= $impl;
            $vars{name} .= $keyed ? '_keyed' : '_flat';
            $vars{name} .= $dedupe ? '_dedupe' : '_dupeok';

            my @params = qw( p_lists limit );
            push @params, 'keyer' if $keyed;
            push @params, 'uniquer' if $dedupe;
            $vars{params} = join ', ', @params;

            $vars{key} = $keyed ? 'callback_value(el, keyer)' : 'key_from_iv(el)';

            my $template_text = <<END_XS;

SV*
[% name %]([% params %])
  SV* p_lists
  IV limit
  [% IF keyed %]SV* keyer[% END %]
  [% IF dedupe %]SV* uniquer[% END %]
CODE:
    AV* lists = (AV*) SvRV(p_lists);
    AV* results = (AV*) sv_2mortal((SV*) newAV());
    IV numlists = av_len(lists) + 1;
    int n;

    [% IF impl == "linear" %]
    lmsxs_ll_ent* heads = NULL;
    [% ELSIF impl == "fib" %]
    struct fibheap* heads = fh_makekeyheap();
    [% END %]

    for (n = 0; n < numlists; n++) {
        AV* list = (AV*) SvRV(*av_fetch(lists, n, 0));
        if (av_len(list) < 0)
            continue;

        SV* el = *av_fetch(list, 0, 0);
        IV key = [% key %];

        [% type %]* ent = [% alloc %];
        [% insert %];
    }

    [% IF dedupe %]HV* seen = (HV*) sv_2mortal((SV*) newHV());
    char uniqbuf[16];[% END %]
    for (n = 0; [% more %] && (!limit || n < limit); ) {
        AV* list;
        [% type %]* ent = [% pop %];

        [% IF dedupe %]
        IV unique = callback_value(ent->sv, uniquer);
        int uniqlen = snprintf(uniqbuf, 16, "%d", unique);
        if (!hv_exists(seen, uniqbuf, uniqlen)) {
            av_push(results, newSVsv(ent->sv));
            n++;
            hv_store(seen, uniqbuf, uniqlen, newSViv(1), 0);
        }
        [% ELSE %]
        av_push(results, newSVsv(ent->sv));
        n++;
        [% END %]

        list = (AV*) SvRV(*av_fetch(lists, ent->list_num, 0));
        if (++ent->list_idx <= av_len(list)) {
            SV* el = *av_fetch(list, ent->list_idx, 0);
            IV key = [% key %];
            ent->sv = el;
            [% IF impl == 'linear' %]ent->key = key;[% END %]
            [% insert %]
        }
        else {
            [% free %](ent);
        }
    }

    while ([% more %])
        [% free %]([% pop %]);
    [% IF impl == "fib" %]fh_deleteheap(heads);[% END %]

    RETVAL = newRV((SV*) results);
OUTPUT:
    RETVAL
END_XS

            my Template $tt = Template->new or die "Couldn't instantiate template: $Template::ERROR";
            my $output;
            $tt->process(\$template_text, \%vars, \$output) or die $tt->error();

            $output =~ s/\n^ +$//mg; # strip whitespace-only lines introduced by TT

            print $output;
        }
    }
}

__DATA__
/*

  DO NOT EDIT!

  This file was automatically generated by gen-XSxs.pl. If that file was not
  included in your distribution, please consult the source repository listed in
  META.yml.

*/

#include "EXTERN.h"
#include "perl.h"
#include "XSUB.h"

#include "ppport.h"

#include "const-c.inc"

#include "fib.h"
#include "list.h"
#include "prio.h"

static
inline
IV key_from_iv(SV* el) {
    if (SvIOK(el))
        return SvIV(el);
    else
        croak("non-integer data encountered");
    return 0;
}

static
inline
IV callback_value(SV* el, SV* callback)
{
    int ret;

    dSP;
    PUSHMARK(SP);
    XPUSHs(el);
    PUTBACK;

    ret = call_sv(callback, G_SCALAR);
    SPAGAIN;

    if (!ret)
        croak("callback did not return a value");

    IV value = POPi;
    PUTBACK;

    return value;
}

MODULE = List::MergeSorted::XS  PACKAGE = List::MergeSorted::XS  PREFIX = l_ms_xs

PROTOTYPES: ENABLE

INCLUDE: const-xs.inc

IV
l_ms_xs_count_elements(p_lists)
  SV* p_lists
CODE:
    AV* lists = (AV*) SvRV(p_lists);
    IV numlists = av_len(lists) + 1;
    int n;
    RETVAL = numlists;

    for (n = 0; n < numlists; n++) {
        AV* list = (AV*) SvRV(*av_fetch(lists, n, 0));
        RETVAL += av_len(list);
    }
OUTPUT:
    RETVAL


XS.pm
package List::MergeSorted::XS;

use 5.008;
use strict;
use warnings;
use Carp;

require Exporter;
use AutoLoader;

our @ISA = qw(Exporter);

our @EXPORT_OK = qw(merge);
our @EXPORT = qw();

our $VERSION = '1.05';

require XSLoader;
XSLoader::load('List::MergeSorted::XS', $VERSION);

use constant {
    PRIO_LINEAR => 0,
    PRIO_FIB    => 1,
    SORT        => 2,
};

sub merge {
    my $lists = shift;
    my %opts = @_;

    # validate inputs
    unless ($lists && ref $lists && ref $lists eq 'ARRAY') {
        die "merge requires an array reference";
    }
    for my $list (@$lists) {
        unless ($list && ref $list && ref $list eq 'ARRAY') {
            die "lists to merge must be arrayrefs";
        }
    }

    my $limit = $opts{limit} || 0;
    die "limit must be positive" if defined $limit && $limit < 0;

    die "key_cb option must be a coderef"
        if defined $opts{key_cb} && ref $opts{key_cb} ne 'CODE';

    die "uniq_cb option must be a coderef"
        if defined $opts{uniq_cb} && ref $opts{uniq_cb} ne 'CODE';

    return [] unless @$lists;

    # pick an algorithm
    my @params = ($lists, $limit, $opts{key_cb}, $opts{uniq_cb});

    if (defined $opts{method}) {
        return _merge($opts{method}, @params);
    }

    if (defined $opts{key_cb}) {
        # linear priority queue is faster until ~100 lists, relatively
        # independent of limit %. sort never wins in keyed mode because of
        # Schwartzian tx overhead

        return scalar @$lists < 100
            ? _merge(PRIO_LINEAR, @params)
            : _merge(PRIO_FIB, @params);
    }
    else {
        # linear always wins with a small number of lists (<100). with more
        # lists, fib wins with low limit, giving way to sort around 25%
        # limit.

        # compute what fraction of the merged set will be returned
        my $total = _count_elements($lists);
        $limit ||= $total;

        if ($limit < 0.05 * $total) {
            return scalar @$lists < 1000
                ? _merge(PRIO_LINEAR, @params)
                : _merge(PRIO_FIB, @params);
        }
        elsif ($limit < 0.25 * $total) {
            return scalar @$lists < 500
                ? _merge(PRIO_LINEAR, @params)
                : _merge(PRIO_FIB, @params)
        }
        elsif ($limit < 0.75 * $total) {
            return scalar @$lists < 100
                ? _merge(PRIO_LINEAR, @params)
                : _merge(SORT, @params)
        }
        else {
            return scalar @$lists < 100
                ? _merge(PRIO_LINEAR, @params)
                : _merge(SORT, @params)
        }
    }
}

# dispatch to appopriate implementation based on algorithm and options
sub _merge {
    my ($method, $lists, $limit, $key_cb, $uniq_cb) = @_;

    if ($method == PRIO_LINEAR) {
        return $key_cb ? $uniq_cb ? _merge_linear_keyed_dedupe($lists, $limit, $key_cb, $uniq_cb)
                                  : _merge_linear_keyed_dupeok($lists, $limit, $key_cb)
                       : $uniq_cb ? _merge_linear_flat_dedupe($lists, $limit, $uniq_cb)
                                  : _merge_linear_flat_dupeok($lists, $limit);
    }
    elsif ($method == PRIO_FIB) {
        return $key_cb ? $uniq_cb ? _merge_fib_keyed_dedupe($lists, $limit, $key_cb, $uniq_cb)
                                  : _merge_fib_keyed_dupeok($lists, $limit, $key_cb)
                       : $uniq_cb ? _merge_fib_flat_dedupe($lists, $limit, $uniq_cb)
                                  : _merge_fib_flat_dupeok($lists, $limit);
    }
    elsif ($method == SORT) {
        return $key_cb ? $uniq_cb ? _merge_sort_keyed_dedupe($lists, $limit, $key_cb, $uniq_cb)
                                  : _merge_sort_keyed_dupeok($lists, $limit, $key_cb)
                       : $uniq_cb ? _merge_sort_flat_dedupe($lists, $limit, $uniq_cb)
                                  : _merge_sort_flat_dupeok($lists, $limit);
    }
    else {
        die "unknown sort method $method requested\n";
    }
}

# concatenate all lists and sort the whole thing. works well when no limit is
# given.

sub _merge_sort_flat_dupeok {
    my ($lists, $limit) = @_;

    my @output = sort {$a <=> $b} map {@$_} @$lists;
    splice @output, $limit if $limit && @output > $limit;
    return \@output;
}

sub _merge_sort_keyed_dupeok {
    my ($lists, $limit, $keyer) = @_;

    # Schwartzian transform is faster than sorting on
    # {$keyer->($a) <=> # $keyer->($b)}, even for degenerately simple case
    # of $keyer = sub { $_[0] }

    my @output =
        map  { $_->[1] }
        sort { $a->[0] <=> $b->[0] }
        map  { [$keyer->($_), $_] }
        map  { @$_ }
        @$lists;

    splice @output, $limit if $limit && @output > $limit;
    return \@output;
}

sub _merge_sort_flat_dedupe {
    my ($lists, $limit, $uniquer) = @_;

    my @merged = sort {$a <=> $b} map {@$_} @$lists;

    my @output;
    my $last_unique = undef;
    for my $element (@merged) {
        my $unique = $uniquer->($element);
        next if defined $last_unique && $unique == $last_unique;
        push @output, $element;
        $last_unique = $unique;
    }
    splice @output, $limit if $limit && @output > $limit;
    return \@output;
}

sub _merge_sort_keyed_dedupe {
    my ($lists, $limit, $keyer, $uniquer) = @_;

    my @merged =
        map  { $_->[1] }
        sort { $a->[0] <=> $b->[0] }
        map  { [$keyer->($_), $_] }
        map  { @$_ }
        @$lists;

    my @output;
    my %seen;
    for my $element (@merged) {
        my $unique = $uniquer->($element);
        next if $seen{$unique}++;
        push @output, $element;
    }

    splice @output, $limit if $limit && @output > $limit;
    return \@output;
}

1;
__END__

=head1 NAME

List::MergeSorted::XS - Fast merger of presorted lists

=head1 SYNOPSIS

  use List::MergeSorted::XS 'merge';

  # merge plain integer lists
  @lists = ([1, 3, 5], [2, 6, 8], [4, 7, 9]);
  merge(\@lists); # [1..9]

  # return only beginning of union
  merge(\@lists, limit => 4); # [1..4]

  # remove duplicates
  @lists = ([1, 2], [0, 2, 3], [3, 4]);
  merge(\@lists, uniq_cb => sub { $_[0] }); # [0..4]

  # merge complicated objects based on accompanying integer keys
  @lists = ([
              [1 => 'x'], [3 => {t => 1}]
            ],
            [
              [2 => bless {}, 'C'], [4 => 5]
            ]);
  $sorted = merge(\@lists, key => sub { $_[0][0] });

=head1 DESCRIPTION

This module takes a set of presorted lists and returns the sorted union of
those lists.

To maximize speed, an appropriate algorithm is chosen heuristically based on
the size of the input data; additionally, efficient C implementations of the
algorithms are used.

=head1 FUNCTIONS

=over 4

=item $merged = merge(\@list_of_lists, %opts)

Computes the sorted union of a set of lists of integers. The first parameter
must be an array reference which itself contains a number of array references.
The result set is returned in an array reference.

The constituent lists must meet these preconditions for correct behavior:

=over 4

=item * either each element of each list is an integer or an integer may be
        computed from the element using the C<key_cb> parameter below

=item * each list is pre-sorted in ascending order

=back

C<merge>'s behavior may be modified by additional options passed after the list:

=over 4

=item * limit

Specifies a maximum number of elements to return. By default all elements are
returned.

=item * key_cb

Specifies a callback routine which will be passed an element of an inner list
through @_. The routine must return the integer value by which the element will be
sorted. In effect, the default callback is C<sub {$_[0]}>. This allows more
complicated structures to be merged.

=item * uniq_cb

Specifies a callback routine which will be passed an element of an inner list
through @_. The routine must return the integer value which identifies the
element in some sense. Elements with the same identity value will not be
duplicated in the output. Elements with the same identity must also have the
same key.

If no uniq_cb is passed, duplicates are allowed in the output.

=item * method

Specifies the algorithm to use to merge the lists. If provided, the value must
be one of the constants listed below under L<ALGORITHM>.

If no B<method> is given, one is chosen automatically based upon the input
data. This is generally recommended.

=back

=back

=head1 NOTES

Only ascending order is supported. To merge lists which are sorted in
descending order, use C<< key_cb => sub { -$_[0] } >>.

=head1 EXPORTS

None by default, C<merge> at request.

=head1 ALGORITHM

The algorithm used to merge the lists is heuristically chosen based on the
number of lists (N), the total number of elements in the lists (M), and the
requested limit (L). (The heuristic constants were determined by analysis of
benchmarks on a 2.5GHz Intel Xeon where all data fit in memory.)

When there are many lists and the element limit is a significant fraction of
the total element count (L/M > 1/4), perl's built-in C<sort> is used to order
the concatenated lists. The time complexity is C<O(M log M)>. Since this method
always processes the full list, it cannot short-circuit in the highly-limited
case (as the priority queue methods do).

When L is a smaller fraction of M, a priority queue is used to track the list
heads. For small N, this is implemented as a linked list kept in sorted order
(using linear-time insertion), yielding time complexity of C<O(L N)>. For large
N, a Fibonacci heap is used, for a time complexity of C<O(L log N)>. The linked
list has less bookkeeping overhead than the heap, so it is more efficient for
fewer lists.

To force a particular implementation, pass the C<method> parameter to C<merge>
with one of these constants:

=over 4

=item * List::MergeSorted::XS::SORT

=item * List::MergeSorted::XS::PRIO_LINEAR

=item * List::MergeSorted::XS::PRIO_FIB

=back

=head1 TODO

* Support comparative orderings, where no mapping from elements to integers
exists but a well-defined ordering exists for which a two-element comparison
callback can be provided.

* Allow modification of the heuristics (perhaps based on local benchmarks).

=head1 SEE ALSO

John-Mark Gurney's Fibonacci heap library L<fib|http://resnet.uoregon.edu/~gurney_j/jmpc/fib.html>

=head1 AUTHOR

Adam Thomason, E<lt>athomason@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2011 by Say Media Inc <cpan@saymedia.com>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.9 or,
at your option, any later version of Perl 5 you may have available.

=cut
